pubtools/pulplib/_impl/client/client.py

```diff
@@ -181,10 +184,87 @@ def search_repository(self, criteria=None):
             response_f, lambda data: self._handle_page(Repository, search, data)
         )
 
+    def get_maintenance_report(self):
+        """Get maintenance report from maintenance repository
+
+        Returns:
+            :class:`~pubtools.pulplib.MaintenceReport` object describes the
+            maintenance status
+
+            None if there's no report in the repository
+        """
+        try:
+            data = self._do_get_maintenance().result()
+        except requests.exceptions.HTTPError:
+            # if the maintenance mode hasn't been used before, then it could return
+            # HTTPError 404.
+            LOG.warning("No maintenance info found")
+            return None
+
+        return MaintenanceReport.from_data(data)
+
+    def set_maintainenance(self, regex, enable=True, **kwargs):
+        """Set repositories to maintenance mode.
+
+        Args:
+            regex (str):
+                A python regular expression pattern used to match repositories.
+
+            enable (bool):
+                If enable, then set the matched repositories to maintenance mode.
+                If not enable, unset the maintenance more for the matched repositories.
+
+            Optional keyword args:
+                owner (str):
+                    Person/party who set/unset maintenance mode.
+                message (str):
+                    Reason why set repositories to maintenance mode, doesn't have
+                    any effect if enable is False.
+
+        Return:
+            A result report in json format (str)
+        """
+        report = self.get_maintenance_report()
```

Comment:
Hm, I don't get the point of making it non blocking. 

For upload or publish, I can get it: user could run many upload and publish tasks in parallel to improve the efficiency. But for set maintenance, with the current mechanism, it's not possible to run several maintenance tasks in parallel -- image if  two tasks download the maintenance file at the same time and update it with different content and upload, the result is the one uploaded first would be overwriten by the latter one. So, if we have to run set maintenance task in serial, why do we need to make it non blocking?

Reply:
OK, there's actually two aspects to this and I was mixing them together wrongly, sorry about that.

Firstly, does it need to be non-blocking: no, strictly speaking it doesn't, for the reason you said.

Secondly, does it need to return a future: yes, I think it does, for the sake of consistency.  Right now, every method which creates Pulp tasks or issues requests to Pulp will return a Future.  If we start making exceptions to that based on implementation details, then the API becomes confusing very fast.  The caller needs to carefully remember the return type of every method and it will seem arbitrary ("getting a repo returns a Future... but getting maintenance report doesn't?")

So I will be happy as long as this returns a Future, it doesn't *have* to be achieved by writing it in non-blocking style.

Reply:
So, for set maintenance, what should be wrapped in the future returned? publish tasks or a result report object?

Reply:
for set maintenance, I plan to do it in blocking-style, since the non-blocking style code I can think of is ugly and hard to read:
`report_ft = self.get_maintenance_report()

        def process_report(report):
            # process report in threads
            if enable:
                # search repositories matched regex
                crit = Criteria.with_field("id", Matcher.regex(regex))
                repos_ft = self.search_repository(crit)
                repo_ids_ft = f_map(repos_ft, lambda repos: [repo.id for repo in repos.as_iter()])
                return f_map(repo_ids_ft, lambda repo_ids: report.add(repo_ids, **kwargs))
            else:
                return f_return(report.remove(regex, **kwargs))

        process_report_ft = f_flat_map(report_ft, process_report)

        repo_ft = self.get_repository("redhat-maintenance")

        # upload updated report to repository and publish
        report_fileobj = f_map(process_report_ft, lambda report: StringIO(report.json()))

        upload_ft = f_flat_map(
            f_sequence(repo_ft, process_report_ft),
            lambda repo_report: repo_report[0].upload_file(repo_report[1])
        )

        publish_ft = f_flat_map(
            f_sequence(repo_ft, upload_ft),
            lambda repo_upload: repo_upload[0].publish()
        )

        return publish_ft`

Reply:
I don't have a strong opinion what should be in the returned future (just not the raw JSON).  A list of tasks created during the process seems the most consistent with other methods right now.

The main use case there would be simply logging which tasks were involved in updating maintenance mode, so you can cross reference with Pulp logs and understand the purpose of each task.